apisix/plugins/auth-hook.lua

```diff
@@ -0,0 +1,337 @@
+--
+-- Licensed to the Apache Software Foundation (ASF) under one or more
+-- contributor license agreements.  See the NOTICE file distributed with
+-- this work for additional information regarding copyright ownership.
+-- The ASF licenses this file to You under the Apache License, Version 2.0
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+
+local core = require("apisix.core")
+local json = require("apisix.core.json")
+local http = require("resty.http")
+local ck = require("resty.cookie")
+local ngx = ngx
+local rawget = rawget
+local rawset = rawset
+local setmetatable = setmetatable
+local string = string
+
+local plugin_name = "auth-hook"
+local hook_lrucache = core.lrucache.new({
+    ttl = 60, count = 1024
+})
+
+local schema = {
+    type = "object",
+    properties = {
+        auth_hook_id = { type = "string", minLength = 1, maxLength = 100, default = "unset" },
+        auth_hook_uri = { type = "string", minLength = 1, maxLength = 4096 },
+        auth_hook_method = {
+            type = "string",
+            default = "GET",
+            enum = { "GET", "POST" },
+        },
+        hook_headers = {
+            type = "array",
+            items = {
+                type = "string",
+                minLength = 1, maxLength = 100
+            },
+            uniqueItems = true
+        },
+        hook_args = {
+            type = "array",
+            items = {
+                type = "string",
+                minLength = 1, maxLength = 100
+            },
+            uniqueItems = true
+        },
+        hook_res_to_headers = {
+            type = "array",
+            items = {
+                type = "string",
+                minLength = 1, maxLength = 100
+            },
+            uniqueItems = true
+        },
+        hook_keepalive = { type = "boolean", default = true },
+        hook_keepalive_timeout = { type = "integer", minimum = 1000, default = 60000 },
+        hook_keepalive_pool = { type = "integer", minimum = 1, default = 5 },
+        hook_res_to_header_prefix = { type = "string", default = "X-", minLength = 1, maxLength = 100 },
+        hook_cache = { type = "boolean", default = false },
+        check_termination = { type = "boolean", default = true },
+    },
+    required = { "auth_hook_uri"},
+}
+
+local _M = {
+    version = 0.1,
+    priority = 1000,
```

Comment:
Because I use proxy-rewrite and then use auth-hook, I found that proxy-rewrite is 1008, so I chose 1000. I think others should also have this requirement. For example, users who have proxy rewrite requirements for path may need to use it. ---auth-hook

Reply:
Because I use proxy-rewrite and then use auth-hook, I found that proxy-rewrite is 1008, so I chose 1000. I think others should also have this requirement. For example, users who have proxy rewrite requirements for path may need to use it.---auth-hook

Reply:
This priority has been taken by node-status (although it is disable by default).

Quote from doc/plugin-develop.md:
> Note : The priority of the new plugin cannot be the same as the priority of any existing plugin.

Reply:
Is it feasible for me to change 1007?

Reply:
Let's do it.